crates/ruff_python_formatter/src/expression/string/docstring.rs

```diff
@@ -395,32 +429,37 @@ impl<'ast, 'buf, 'fmt, 'src> DocstringLinePrinter<'ast, 'buf, 'fmt, 'src> {
     /// routine is silent about it. So from the user's perspective, this will
     /// fail silently. Ideally, this would at least emit a warning message,
     /// but at time of writing, it wasn't clear to me how to best do that.
-    ///
-    /// # Panics
-    ///
-    /// This panics when the given slice is empty.
     fn format(
         &mut self,
         code: &[CodeExampleLine<'_>],
     ) -> FormatResult<Option<Vec<OutputDocstringLine<'static>>>> {
         use ruff_python_parser::AsMode;
 
-        let offset = code
-            .get(0)
-            .expect("code blob must be non-empty")
-            .original
-            .offset;
-        let last_line_is_last = code
-            .last()
-            .expect("code blob must be non-empty")
-            .original
-            .is_last();
+        let (Some(unformatted_first), Some(unformatted_last)) = (code.first(), code.last()) else {
+            return Ok(None);
+        };
         let codeblob = code
             .iter()
             .map(|line| line.code)
             .collect::<Vec<&str>>()
             .join("\n");
-        let printed = match docstring_format_source(self.f.options(), self.quote_style, &codeblob) {
+        let options = self
+            .f
+            .options()
+            .clone()
+            // It's perhaps a little odd to be hard-coding the indent
+            // style here, but I believe it is necessary as a result
+            // of the whitespace normalization otherwise done in
+            // docstrings. Namely, tabs are rewritten with ASCII
+            // spaces. If code examples in docstrings are formatted
+            // with tabs and those tabs end up getting rewritten, this
+            // winds up screwing with the indentation in ways that
+            // results in formatting no longer being idempotent. Since
+            // tabs will get erased anyway, we just clobber them here
+            // instead of later, and as a result, get more consistent
+            // results.
```

Comment:
I would be interested in your opinion on https://github.com/astral-sh/ruff/issues/8430 after dealing with tab/space indentation in code examples.

Reply:
Aye. I left a comment. To me it feels like one of the issues here is hard-coding tabwidth=8. So long as that's there, it will be tough for us to cleanly solve the problem.